### PR TITLE
Fluent API RecipeBuilder syntax

### PR DESCRIPTION
**Pull request by webmilio**
### What is the new feature?
Recipe building using a fluent api style syntax.

### Why should this be part of tModLoader?
Because the current way recipes are added is a bit lacking.

For more information, check out [this issue](https://github.com/tModLoader/tModLoader/issues/847).

### CODE REVIEW

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+			_recipe = new ModRecipe(mod);
+			_recipe.SetResult(mod, itemName, stack);
+		}
+
+		/// <summary>Creates a new instance with no ingredients with the given item and stack as a result.</summary>
+		/// <param name="item"></param>
+		/// <param name="stack"></param>
+		/// <seealso cref="Requires(int,int)"/>
+		/// <seealso cref="Build"/>
+		public RecipeBuilder(ModItem item, int stack = 1)
```
Comment (Jofairden) [inline patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:49]: I would get rid of this, it's a bit dangerous to just 'inherit' the mod of the item. You will be adding a recipe 'as another mod' if the item comes from another mod.

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+		/// Adds the specified ingredients to this recipe with the given item types.
+		/// Ex.: 
+		/// <example>recipe.AddIngredient(ItemID.IronAxe)</example>
+		/// </summary>
+		/// <param name="type1">The first item type.</param>
+		/// <param name="type2">The second item type.</param>
+		/// <param name="type3">The third item type.</param>
+		/// <param name="types">The remaining item types.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires(int type1, int type2, int type3, params int[] types)
```
Comment (Jofairden) [inline patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:92]: Why type1 through 3 and then params types? Just do params types.

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+		/// <summary>Adds an ingredient to this recipe with the given item name from the given mod, and with the given stack stack. If the mod parameter is null, then it will automatically use an item from the mod creating this recipe.</summary>
+		/// <param name="mod">The mod.</param>
+		/// <param name="itemName">Name of the item.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		/// <exception cref="RecipeException">The item " + itemName + " does not exist in mod " + mod.Name + ". If you are trying to use a vanilla item, try removing the first argument.</exception>
+		public RecipeBuilder Requires(Mod mod, string itemName, int stack = 1)
+		{
+			_recipe.AddIngredient(mod ?? _recipe.mod, itemName, stack);
```
Comment (Jofairden) [inline patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:115]: Why the null coalescing? The player should pass mod, if it is incorrect there should be an error

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+			need(_recipe);
+
+			return this;
+		}
+
+
+		/// <summary>Adds this recipe to the game. Call this after you have finished setting the result, ingredients, etc.</summary>
+		/// <returns></returns>
+		/// <exception cref="RecipeException">A recipe without any result has been added.</exception>
+		public void Build() => _recipe.AddRecipe();
```
Comment (Jofairden) [inline patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:182]: You should have a 'finalized' or 'added' bool on this builder and check against it in any of the method calls, the builder is 'done' once Build() is called.

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+{
+	public class RecipeBuilder
+	{
+		private readonly ModRecipe _recipe;
+
+
+		/// <summary>Creates a new instance with no ingredients and no result.</summary>
+		/// <seealso cref="Requires(int,int)"/>
+		/// <seealso cref="Build"/>
+		public RecipeBuilder()
```
Comment (Jofairden) [inline patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:16]: It also doesn't initialize the recipe, remove this.

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs (hunk 3 above)
Comment (webmilio) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:115]: As per the TML documentation, if the passed mod is null, it'll use the one who created the recipe.

See http://tmodloader.github.io/tModLoader/html/class_terraria_1_1_mod_loader_1_1_mod_recipe.html#a7c390feb8cf0a3212de31b53ad5bcfc2.

I could let `AddIngredient` do the work if thats what it already does I suppose.

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs (hunk 1 above)
Comment (webmilio) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:49]: This is the equivalent of `ModRecipe.SetResult(ModItem, int)`, which is pretty standard in most mods adding ModItems with recipes. I don't think this should be removed, but maybe fixed ? Adding a mod parameter perhaps ?

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs (hunk 4 above)
Comment (webmilio) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:182]: Thats a lot of hand-holding. Do we need to throw an error every time as well ?

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs (hunk 5 above)
Comment (webmilio) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:16]: Will do.

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs (hunk 2 above)
Comment (webmilio) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:92]: I did as was discussed on this [comment](https://github.com/tModLoader/tModLoader/issues/847#issuecomment-621730542), but I will change it back to `Requires(params int[])`.

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs (hunk 3 above)
Comment (Mirsario) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:103]: Strings should only be usable for accessing other mods' content in such a manual context. Why is there even a string overload if there isn't one with a generic ModItem parameter?

Comment (webmilio) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:103]: I'm not entirely sure what you mean. I followed the existing structure that ModRecipe provides. Do you mean to say we should have a `Require<T>(...) where T : ModItem` ?

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs (hunk 1 above)
Comment (Jofairden) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:49]: Ok if it's the same as that one, then it's fine

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+		public RecipeBuilder RequiresHoney() => NeedVar(recipe => recipe.needHoney = true);
+
+		public RecipeBuilder RequiresWater() => NeedVar(recipe => recipe.needWater = true);
+
+		public RecipeBuilder RequiresSnowBiome() => NeedVar(recipe => recipe.needSnowBiome = true);
+
+
+		private RecipeBuilder NeedVar(Action<ModRecipe> need)
+		{
+			need(_recipe);
```
Comment (Jofairden) [inline patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:179]: What is need ?

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs (hunk 2 above)
Comment (Jofairden) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:92]: You put this back, by accident?

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs [hunk 7]
```diff
@@ -100,11 +112,17 @@ public RecipeBuilder Requires(params int[] types)
...
+			_recipe.AddIngredient(mod, itemName, stack);
 
 			return this;
 		}
 
+		/// <summary>Adds an ingredient to this recipe of the given type and stack size.</summary>
+		/// <typeparam name="T">The type.</typeparam>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires<T>(int stack = 1) where T : ModItem => Requires(ModContent.ItemType<T>(), stack);
```
Comment (Jofairden) [inline patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:124]: If we're on this route, you could make a `Requires(params Type[] types)` so you can do `Requires(typeof(MyItem))`
But it gets pretty verbose at that point, `Requires<T>` is fine

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs (hunk 6 above)
Comment (webmilio) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:179]: Simple method I made that sets a flag to true and returns RecipeBuilder.

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs (hunk 2 above)
Comment (webmilio) [reply patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:92]: Yea, I'll remove it again. Seems I forgot to fetch before pulling.

File: patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+		public RecipeBuilder At(params int[] tileTypes)
+		{
+			for (int i = 0; i < tileTypes.Length; i++)
+				_recipe.AddTile(tileTypes[i]);
+
+			return this;
+		}
+
+
+		public RecipeBuilder RequiresLava() => NeedVar(recipe => recipe.needLava = true);
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs:160]: no need for an Action and a common method here. Yes it lets you use expression methods. No it makes it harder to read.
```
public RecipeBuilder RequiresLava() {
    _recipe.needLave = true;
    return this;
}
```

**Comment by Mirsario** [comment]
I've realized that this is completely unfit for custom ModRecipe types. We'll need to look into redesigning recipes, because extending the ModRecipe class currently looks pretty ugly, and still will with RecipeBuilder. Have a look: https://paste.mod.gg/rupotepaqi.cs

**Comment by Chicken-Bones** [comment]
ModRecipe should be a builder itself. There's really no need for a separate builder class. Just make all the existing functions `return this;`

**Comment by webmilio** [comment]
> I've realized that this is completely unfit for custom ModRecipe types. We'll need to look into redesigning recipes for OOP, because extending the ModRecipe class currently looks pretty ugly, and still will with RecipeBuilder. Have a look: https://paste.mod.gg/esetilitug.cs

What exactly does this show ? I don't see you using the RecipeBuilder for comparaison.

<br>

> ModRecipe should be a builder itself. There's really no need for a separate builder class. Just make all the existing functions `return this;`

I agree, I wasn't too sure if I could modify ModRecipe. I'll be unable to change anything for the next week and a half sadly.

**Comment by Chicken-Bones** [comment]
This might've been too hasty. I think there was some good syntax discussion we could use for improvement, and changing signatures of current `ModRecipe` methods is a breaking change.

**Comment by Jofairden** [comment]
It was, I kind of assumed it'd work because all I was thinking of was "if you don't use the return value right now that doesn't matter", though that doesn't matter for already compiled mods >.>

**Comment by Itorius** [comment]
Fluent API was already merged into 1.4.

**Pull request closed without merging**